walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,87 @@
 
 package com.google.sps;
 
+import java.sql.Time;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  private static final Collection<Event> NO_EVENTS = Collections.emptySet();
+  private static final Collection<String> NO_ATTENDEES = Collections.emptySet();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> ranges = new ArrayList<>();
+    if((events == NO_EVENTS&& request.getDuration()<=TimeRange.WHOLE_DAY.duration()) 
+      || request.getAttendees() == NO_ATTENDEES){
+      ranges.add(TimeRange.WHOLE_DAY);
+      return ranges;
+    }else if(request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return ranges;
+    }
+    Collection<String> attendees = request.getAttendees();
+    for(Event event: events){
+      Collection<String> eAttendees = event.getAttendees();
+      for(String attendee: attendees){
+        if(eAttendees.contains(attendee)||eAttendees.isEmpty()){
+          ranges.add(event.getWhen());
+        }
+      }
+    }
+    if(ranges.size()==0){
+      ranges.add(TimeRange.WHOLE_DAY);
+      return ranges;
+    }
+    Collection<TimeRange> overlapedRanges = new ArrayList<>();
+    if(ranges.size()>1){
+      ranges.sort(TimeRange.ORDER_BY_START);
+      TimeRange r1 = ranges.get(0);
+      for(int i = 1; i < ranges.size();i++){
+        TimeRange r2 = ranges.get(i);
+        if(r1.overlaps(r2)){
```

Comment:
Fundamentally, this strategy of looking only one ahead and then adding to overlaped Ranges is not going to work. What if you have

r1 = [0, 5)
r2 = [3, 9);
r3 = [7, 15)

Then, you should have only a single interval in overlapedRanges containing [0, 15), but your code will end up putting two ranges I think.

I would recommend an alternate organization:

List overlapedRanges = new ArrayList<>();
for(int i =0; i < num_ranges; i++) {
Range currentUnion = ranges.get(i);
Range next =ranges.get(i+1);
while(current_union.contains(next)) {
currentUnion = currentUnion.combine(next);
i++;
}
overlapedRanges.add(currentUnion);
}